flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeFsUploader.java

```diff
@@ -168,7 +168,7 @@ private UploadTasksResult upload(Path path, Collection<UploadTask> tasks) throws
                 wrappedStreamClosed = true;
             }
         } finally {
-            if (!wrappedStreamClosed) {
+            if (!wrappedStreamClosed || compression) {
                 fsStream.close();
             }
```

Comment:
The logic becomes a bit tricky IMO.
I think we can simplify it by:
1. wrapping the stream we want to close 
2. in wrapper, making sure it is closed at most once
3. close the wrapper using `try-finally`

And we also get rid of `wrappedStreamClosed`.

Something like this:

```
try (FSDataOutputStream fsStream =
    wrapWithCloseOnce(fileSystem.create(path, NO_OVERWRITE))) {
    try (OutputStreamWithPos stream = wrap(fsStream)) {
        ...
    }
}

```
```
private static FSDataOutputStream wrapWithCloseOnce(FSDataOutputStream stream) {
    return new FSDataOutputStream() {
        private boolean closed;

        @Override
        public void close() throws IOException {
            if (!closed) {
                closed = true;
                stream.close();
            }
        }

        ...
    };
}
```
WDYT?

Reply:
Nice suggestion.

Reply:
An alternative could be to remove
```
OutputStream compressed =
        compression ? instance.decorateWithCompression(fsStream) : fsStream;
```

Could you check if that causes the build to fail?

Reply:
An alternative could be to remove
```
OutputStream compressed =
        compression ? instance.decorateWithCompression(fsStream) : fsStream;
```

Could you check if that causes the build to fail?

Reply:
I have tried to remove and rerun it, it also cannot work. It's not the reason.
Actually, I found the the case will always fail whether enabling compression or not.
What confused me is that the case will always fail only if I warp the stream even if there is no other logic.
And It just keep running until timeout and don't thrown an exception.

Reply:
I mean that with [this change](https://github.com/apache/flink/pull/20306#discussion_r930161714), neither wrapping, nor additional flag check (`wrappedStreamClosed`) would be required.

Currently, the composition of streams looks like this:
1. `OutputStreamWithPos` <-- closed by the inner `try` block
1. `BufferedOutputStream`
1. `SnappyFramedOutputStream` or `UncompressedStreamCompressionDecorator`
1. maybe `NonClosingOutputStreamDecorator` <- doesn't proxy `close()`
1. `FSDataOutputStream` <-- closed by the outer `try` block

Where (4) `NonClosingOutputStreamDecorator` is conditional in `wrap()`.
But that condition is unnecessary because `StreamCompressionDecorator` instance is already choosen in (3).

So if `decorateWithCompression` would **always** add `NonClosingOutputStreamDecorator` then it should solve the problem.

WDYT?

Reply:
I think your analysis is right.
I have modified it and found it works.
Thanks a lot!